### PR TITLE
Refactor method for list functions and force refresh during retry

### PR DESCRIPTION
**Pull request by Flanker32**
Refactor method for list functions and force refresh during retry

### CODE REVIEW

**Review by andxu — DISMISSED** [review]

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java [hunk 1]
```diff
@@ -113,9 +119,10 @@
     private static final String UPDATE_FUNCTION_DONE = "Successfully updated Function App %s.";
     private static final String NO_ARTIFACT_FOUNDED = "Failed to find function artifact '%s.jar' in folder '%s', please re-package the project and try again.";
     private static final String LOCAL_SETTINGS_FILE = "local.settings.json";
-    private static final int LIST_TRIGGERS_MAX_RETRY = 3;
+    private static final int LIST_TRIGGERS_MAX_RETRY = 4;
```
Comment (andxu) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java:122]: odd number 4

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java [hunk 2]
```diff
@@ -440,24 +447,16 @@ protected void listHTTPTriggerUrls(IFunctionApp target) {
...
-                // swallow service exception while list triggers
-            }
-            try {
-                Thread.sleep(LIST_TRIGGERS_RETRY_PERIOD_IN_SECONDS * 1000);
-            } catch (InterruptedException e) {
-                // swallow interrupted exception
-            }
-        }
-        throw new AzureToolkitRuntimeException(NO_TRIGGERS_FOUNDED);
+        final AtomicInteger count = new AtomicInteger(0);
```
Comment (wangmingliang-ms) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java:450]: use `AtomicBoolean` if needed, and maybe `boolean` is enough

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java [hunk 3]
```diff
@@ -440,24 +447,16 @@ protected void listHTTPTriggerUrls(IFunctionApp target) {
...
+        return Mono.fromCallable(() -> {
+            final AzureString message = count.getAndAdd(1) == 0 ?
+                    AzureString.fromString(SYNCING_TRIGGERS) : AzureString.format(SYNCING_TRIGGERS_WITH_RETRY, count.get(), LIST_TRIGGERS_MAX_RETRY);
+            AzureMessager.getMessager().info(message);
+            return Optional.ofNullable(functionApp.listFunctions(true))
+                    .filter(CollectionUtils::isNotEmpty)
+                    .orElseThrow(() -> new AzureToolkitRuntimeException(NO_TRIGGERS_FOUNDED));
+        }).subscribeOn(Schedulers.boundedElastic())
+                .retryWhen(Retry.backoff(LIST_TRIGGERS_MAX_RETRY - 1, Duration.ofSeconds(LIST_TRIGGERS_RETRY_PERIOD_IN_SECONDS))).block();
```
Comment (wangmingliang-ms) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java:459]: will `block()` throw exception `NO_TRIGGERS_FOUNDED` if the last try still fails?

Comment (Flanker32) [reply azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java:459]: Will throw `reactor.core.Exceptions$RetryExhaustedException` while no triggers found was the cause. And this exception will be caught by called and show warning for users

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java (hunk 2 above)
Comment (Flanker32) [reply azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java:450]: Need a final indicator in lambda, so use `AtomicInteger` here

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java (hunk 1 above)
Comment (Flanker32) [reply azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java:122]: Refers https://github.com/microsoft/vscode-azurefunctions/blob/main/src/commands/deploy/notifyDeployComplete.ts#L40, 1 normal call + 3 retry

**Review by andxu — APPROVED** [review]

**Pull request merged**